src/coap_mbedtls.c

```diff
@@ -0,0 +1,1830 @@
+  }
+
+  /*
+   * Verify Peer.
+   *  Need to do all checking, even if setup_data->verify_peer_cert is not set
+   */
+  mbedtls_ssl_conf_verify(&m_env->conf,
+                          cert_verify_callback_mbedtls, c_session);
+
+  return 0;
+}
+
+#if defined(MBEDTLS_SSL_SRV_C)
+/*
+ * PKI SNI callback.
+ */
+static int
+pki_sni_callback(void *p_info, mbedtls_ssl_context *ssl,
+             const unsigned char *uname, size_t name_len)
+{
+  unsigned int i;
+  coap_dtls_pki_t sni_setup_data;
+  coap_session_t *c_session = (coap_session_t *)p_info;
+  coap_mbedtls_env_t *m_env = (coap_mbedtls_env_t *)c_session->tls;
+  coap_mbedtls_context_t *m_context =
+           (coap_mbedtls_context_t *)c_session->context->dtls_context;
+  int ret = 0;
+  char *name;
+
+  name = mbedtls_malloc(name_len+1);
+  if (!name)
+    return -1;
+
+  memcpy(name, uname, name_len);
+  name[name_len] = '\000';
+
+  /* Is this a cached entry? */
+  for (i = 0; i < m_context->pki_sni_count; i++) {
+    if (strcasecmp(name, m_context->pki_sni_entry_list[i].sni) == 0) {
+      break;
+    }
+  }
+  if (i == m_context->pki_sni_count) {
+    /*
+     * New PKI SNI request
+     */
+    coap_dtls_key_t *new_entry;
+
+    new_entry =
+      m_context->setup_data.validate_sni_call_back(name,
+                                 m_context->setup_data.sni_call_back_arg);
+    if (!new_entry) {
+      ret = -1;
+      mbedtls_free(name);
+      goto end;
+    }
+
+    m_context->pki_sni_entry_list =
+             mbedtls_realloc(m_context->pki_sni_entry_list,
+                                   (i+1)*sizeof(pki_sni_entry));
+    m_context->pki_sni_entry_list[i].sni = name;
+    m_context->pki_sni_entry_list[i].pki_key = *new_entry;
+    sni_setup_data = m_context->setup_data;
+    sni_setup_data.pki_key = *new_entry;
+    if ((ret = setup_pki_credentials(&m_context->pki_sni_entry_list[i].cacert,
+                         &m_context->pki_sni_entry_list[i].public_cert,
+                         &m_context->pki_sni_entry_list[i].private_key,
+                         m_env,
+                         m_context,
+                         c_session,
+                         &sni_setup_data, COAP_DTLS_ROLE_SERVER)) < 0) {
+      ret = -1;
+      mbedtls_free(name);
+      goto end;
+    }
+    m_context->pki_sni_count++;
+  }
+  else {
+    mbedtls_free(name);
+  }
+
+end:
+  if (ret != -1) {
+    mbedtls_ssl_set_hs_ca_chain(ssl, &m_context->pki_sni_entry_list[i].cacert,
+                                NULL);
+    return mbedtls_ssl_set_hs_own_cert(ssl,
+                                &m_context->pki_sni_entry_list[i].public_cert,
+                                &m_context->pki_sni_entry_list[i].private_key);
+  }
+  return ret;
+}
+#endif /* MBEDTLS_SSL_SRV_C */
+
+/*
+ * PSK SNI callback.
+ */
+static int
+psk_sni_callback(void *p_info, mbedtls_ssl_context *ssl,
+             const unsigned char *uname, size_t name_len)
+{
+  unsigned int i;
+  coap_dtls_spsk_t sni_setup_data;
```

Comment:
This variable seems to have become obsolete:

```
src/coap_mbedtls.c: In function ‘psk_sni_callback’:
src/coap_mbedtls.c:778:20: warning: variable ‘sni_setup_data’ set but not used [-Wunused-but-set-variable]
  778 |   coap_dtls_spsk_t sni_setup_data;
      |                    ^~~~~~~~~~~~~~
```

Reply:
Redundant code now removed.